project_sla/analytic_account.py

```diff
@@ -38,31 +38,36 @@ def _reapply_sla(self, cr, uid, ids, recalc_closed=False, context=None):
         The ``recalc_closed`` flag allows to also recompute closed documents.
         """
         ctrl_obj = self.pool.get('project.sla.control')
+        sla_obj = self.pool.get('project.sla')
+
+        context = {} if context is None else context.copy()
 
         exclude_states = ['cancelled', 'cancel']
         if not recalc_closed:
             exclude_states += ['done']
 
-        for contract in self.browse(cr, uid, ids, context=context):
+        # If this is called from SLA model, than we could reduce number of
+        # documents that need recalculation
+        if context.get('sla_ids', False):
+            sla_ids = context['sla_ids']
+        else:
+            sla_ids = sla_obj.search(cr, uid, [('analytic_ids', 'in', ids)],
+                                     context=context)
+
+        sla_recs = sla_obj.browse(cr, uid, sla_ids, context=context)
+        for m_name in set([sla.control_model for sla in sla_recs]):
             # for each contract, and for each model under SLA control ...
-            for m_name in set([sla.control_model for sla in contract.sla_ids]):
-                model = self.pool.get(m_name)
-                doc_ids = []
-                if 'analytic_account_id' in model._columns:
-                    doc_ids += model.search(
-                        cr, uid,
-                        [('analytic_account_id', '=', contract.id),
-                         ('state', 'not in', exclude_states)],
-                        context=context)
-                if 'project_id' in model._columns:
-                    doc_ids += model.search(
-                        cr, uid,
-                        [('project_id.analytic_account_id', '=', contract.id),
-                         ('state', 'not in', exclude_states)],
-                        context=context)
-                if doc_ids:
-                    docs = model.browse(cr, uid, doc_ids, context=context)
-                    ctrl_obj.store_sla_control(cr, uid, docs, context=context)
+            model = self.pool.get(m_name)
+            domain = [('state', 'not in', exclude_states)]
+            if 'analytic_account_id' in model._columns:
+                domain += [('analytic_account_id', 'in', ids)]
+            if 'project_id' in model._columns:
+                domain += [('project_id.analytic_account_id', 'in', ids)]
```

Comment:
@dreispt What do You think, is there should be 'or' or 'and' logic in domain, if there are both fields ('analytic_account_id' and 'project_id') present in model?